Promotions edit: Add enable-able conditions

coderkevin opened the pull request:
This adds enable-able conditions and some supporting code including a
DateField, NumberField, and some modifications to the way FormField
works that allows pass-thru props, and default values.

To Test:
1. `npm start`
2. `http://calypso.localhost:3000/store/promotion/<site url>`
3. For a coupon type, try out each condition to ensure you can enter data and it stays (can also check in redux devtools to verify)
4. For a product sale, try out the start date and end date fields.

![new_promotion_ _coderkevin_ _wordpress_com](https://user-images.githubusercontent.com/945228/32132256-514652e0-bb86-11e7-8a32-a7005e2e76ed.png)

kellychoffman commented:
I updated the copy, but could you align the inputs to the labels? IE: 
![screen shot 2017-10-30 at 11 03 35 am](https://user-images.githubusercontent.com/5835847/32184503-09f4a882-bd62-11e7-98e6-d941ccbfee5e.png)

coderkevin commented:
>  I could chose an end date that was before the start date. 

I don't think this is going to have a super easy solution, so I'm preemptively filing an issue to track this and fix it in a future PR: #19313

coderkevin commented:
>I updated the copy, but could you align the inputs to the labels?

Added margin in 193c780

coderkevin commented on client/extensions/woocommerce/app/promotions/promotion-models.js:
```diff
@@ -75,6 +96,55 @@ const productSaleModel = {
+		},
+		maximumAmount: {
+			component: CurrencyField,
+			labelText: translate( 'The total discount cannot exceed a certain amount' ),
```
@kellychoffman I don't think this is the right wording. I think this is actually a maximum spend of sorts. I don't know what the actual code does, but I think it probably limits the maximum benefit of the coupon to this spending limit. Here's what wp-admin says about it:

![screen shot 2017-10-30 at 10 40 18 pm](https://user-images.githubusercontent.com/945228/32206508-83fbc23a-bdc3-11e7-9847-197f366e521d.png)

jameskoster commented:
The margin you added if affecting the other cards;

![margin](https://cldup.com/kVDtUY0I6m-2000x2000.png)

There doesn't seem to be a unique class per card though so I couldn't fix.

coderkevin commented:
> The margin you added if affecting the other cards;
Oops, I'll fix.

jameskoster commented:
@coderkevin I'd like to tweak the condition list styling a bit, but it's difficult without a class on the card. Would it be OK to add one please? `.promotions__promotion-conditions` or something.

coderkevin commented:
@jameskoster I added the ability to have css classes in the model and I added two such classes: `promotions__promotion-form-card-primary` and `promotions__promotion-form-card-conditions`. Let me know if you need anything else here.

allendav commented on client/extensions/woocommerce/app/promotions/fields/README.md:
```diff
@@ -33,6 +33,10 @@ Translated text to be shown below the field children.
 
 If true, show a "Required" annotation next to the label.
 
+### `isEnableable` (optional)
```
I love this word :P

allendav commented on client/extensions/woocommerce/app/promotions/fields/form-field.js:
```diff
@@ -16,21 +18,50 @@ const FormField = ( {
+
+	const childrenClassNames = classNames(
+		'fields__fieldset-children',
+		{ 'fields__fieldset-children-enableable': enableCheckbox }
```
Optional. Replacing enableCheckbox with isEnableable would seem less odd to me. :oldmanyellsatcloud:

coderkevin replied:
In this case, it's actually using the enable checkbox as a binary check to see if it should add the class.

allendav commented on client/extensions/woocommerce/app/promotions/promotion-models.js:
```diff
@@ -75,6 +97,57 @@ const productSaleModel = {
+		},
+		usageLimit: {
+			component: NumberField,
+			labelText: translate( 'Limit number of times this promotion can be used in total' ),
```
Nit. Would `Limit **the** number of times this promotion...` be better grammar?

allendav commented on client/extensions/woocommerce/app/promotions/promotion-models.js:
```diff
@@ -75,6 +97,57 @@ const productSaleModel = {
+		},
+		usageLimitPerUser: {
+			component: NumberField,
+			labelText: translate( 'Limit total times each customer can use this promotion' ),
```
Nit. Would

Limit **the** total **number** of times each...

be better grammar?

coderkevin commented:
@jameskoster Would you like to address the comments about the copy above from @allendav?

I'm going to go ahead and merge this, but feel free to discuss still. We can always open another PR.
